File: frontend/lib/api/backtest.ts
import { apiPost } from "./client";

export interface BacktestRequest {
  scheme_codes: number[];
  weights: Record<number, number>;
  mode: "Lump Sum" | "SIP (Monthly)";
  lump_sum_amount: number;
  sip_amount: number;
  rebalance_freq: "None" | "Monthly" | "Quarterly" | "Annually";
  start_date: string;
  end_date: string;
}

export interface BacktestResultRow {
  [key: string]: unknown;
  nav_date: string;
  portfolio_value: number;
  total_invested: number;
}

export interface TwrMetrics {
  cagr_pct: number | null;
  sharpe_ratio: number | null;
  max_drawdown_pct: number | null;
  vol_annualized_pct: number | null;
  [key: string]: unknown;
}

export interface BacktestResult {
  error?: string;
  missing_codes?: number[];
  df_result?: BacktestResultRow[];
  twr_metrics?: TwrMetrics;
  final_value?: number;
  total_invested?: number;
  n_contributions?: number;
  absolute_gain?: number;
  absolute_return_pct?: number | null;
  money_weighted_xirr_pct?: number | null;
  normalized_weights?: Record<number, number>;
  actual_start?: string;
}

export const runBacktest = (req: BacktestRequest) => apiPost<BacktestResult>("/api/backtest", req);

File: backend/app/routers/backtest.py
import datetime
from typing import Dict, List

from fastapi import APIRouter
from pydantic import BaseModel

from app.core.serialize import df_to_records, sanitize_floats
from app.services import backtest as backtest_service

router = APIRouter(prefix="/api/backtest", tags=["backtest"])


class BacktestRequest(BaseModel):
    scheme_codes: List[int]
    weights: Dict[int, float]  # raw (not necessarily normalized) weights, keyed by scheme_code
    mode: str  # "Lump Sum" or "SIP (Monthly)"
    lump_sum_amount: float = 0.0
    sip_amount: float = 0.0
    rebalance_freq: str = "None"
    start_date: datetime.date
    end_date: datetime.date


@router.post("")
def run_backtest(req: BacktestRequest) -> dict:
    result = backtest_service.run_portfolio_backtest(
        scheme_codes=req.scheme_codes,
        weights=req.weights,
        mode=req.mode,
        lump_sum_amount=req.lump_sum_amount,
        sip_amount=req.sip_amount,
        rebalance_freq=req.rebalance_freq,
        start_date=req.start_date,
        end_date=req.end_date,
    )
    if "error" in result:
        return {"error": result["error"], "missing_codes": result.get("missing_codes", [])}

    df_result = result.pop("df_result")
    return sanitize_floats(
        {
            **result,
            "df_result": df_to_records(df_result),
        }
    )
